Mathlib/Data/Nat/Prime/MersenneFermat.lean

```diff
@@ -0,0 +1,188 @@
+/-
+Copyright (c) 2024 Ralf Stephan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Ralf Stephan, Thomas Browning
+-/
+import Mathlib.Data.Nat.Prime.Basic
+import Mathlib.Data.Nat.Factorization.Basic
+import Mathlib.Data.ZMod.Basic
+import Mathlib.FieldTheory.Finite.Basic
+import Mathlib.NumberTheory.LegendreSymbol.QuadraticReciprocity
+import Mathlib.NumberTheory.LucasPrimality
+
+/-!
+# Lemmas about Fermat numbers
+-/
+
+open ZMod Nat
+
+/-- Prime `a ^ n + 1` implies `n` is a power of two (Fermat primes). -/
+theorem pow_of_pow_add_prime {a n : ℕ} (ha : 1 < a) (hn : n ≠ 0) (hP : (a ^ n + 1).Prime) :
+    ∃ m : ℕ, n = 2 ^ m := by
+  obtain ⟨k, m, hm, rfl⟩ := exists_eq_two_pow_mul_odd hn
+  rw [pow_mul] at hP
+  use k
+  replace ha : 1 < a ^ 2 ^ k := one_lt_pow (pow_ne_zero k two_ne_zero) ha
+  let h := hm.nat_add_dvd_pow_add_pow (a ^ 2 ^ k) 1
+  rw [one_pow, hP.dvd_iff_eq (Nat.lt_add_right 1 ha).ne', add_left_inj, pow_eq_self_iff ha] at h
+  rw [h, mul_one]
+
+/-- `Fₙ = 2^(2^n)+1` is prime only if `3^(2^(2^n-1)) = -1 mod Fₙ` (Pépin's test). -/
+lemma pepin_primality (n : ℕ) (h : 3 ^ (2 ^ (2 ^ n - 1)) = (-1 : ZMod (2 ^ (2 ^ n) + 1))) :
+    (2 ^ (2 ^ n) + 1).Prime := by
+  have hneg1ne1: (-1 : ZMod (2 ^ (2 ^ n) + 1)) ≠ (1 : ZMod (2 ^ (2 ^ n) + 1)) := by
+    by_contra h'
+    let h := (ZMod.neg_eq_self_iff (1 : ZMod (2 ^ (2 ^ n) + 1))).mp h'
+    rcases h with h | h
+    · absurd h
+      rw [Fin.one_eq_zero_iff, ← ne_eq, succ_ne_succ]
+      exact NeZero.ne (2 ^ 2 ^ n)
+    · absurd h
+      have : Fact (1 < 2 ^ (2 ^ n) + 1) := by
+        simp only [lt_add_iff_pos_left, ofNat_pos, pow_pos]
+        exact { out := trivial }
+      rw [val_one]
+      norm_num
+      simp only [reduceEqDiff]
+      rw [pow_eq_one_iff (x := 2) (n := 2 ^ n)]
+      · exact succ_succ_ne_one 0
+      · exact Ne.symm (NeZero.ne' (2 ^ n))
+  apply lucas_primality (p := 2 ^ (2 ^ n) + 1) (a := 3)
+  · norm_num
+    apply Mathlib.Tactic.Ring.pow_nat _ h neg_one_sq
+    rw [mul_comm, ← Nat.pow_add' 2 (2 ^ n - 1) 1]
+    norm_num
+    rw [Nat.sub_add_cancel Nat.one_le_two_pow]
+  · norm_num
+    intro p H' H''
+    rw [(Nat.prime_dvd_prime_iff_eq H' prime_two).mp ((Nat.Prime.dvd_of_dvd_pow H') H'')]
+    have : 2 ^ (2 ^ n) / 2 = 2 ^ (2 ^ n - 1) := by
+      rw [← Nat.mul_right_inj (a := 2), ← Nat.pow_add' 2 (2 ^ n - 1) 1,
+        mul_comm, Nat.div_mul_cancel]
+      · norm_num
+        rw [Nat.sub_add_cancel Nat.one_le_two_pow]
+      · exact dvd_pow_self 2 (Ne.symm (NeZero.ne' (2 ^ n)))
+      · exact Ne.symm (zero_ne_add_one 1)
+    rw [this, h]
+    exact hneg1ne1
+
+/- Prime factors of `Fₙ = 2 ^ (2 ^ n) + 1`, `1 < n`, are of form `k * 2 ^ (n + 2) + 1`. -/
+lemma fermat_primeFactors_one_lt (n p : ℕ) (hn : 1 < n) (hP : p.Prime) (hp' : p ≠ 2)
+    (hpdvd : p ∣ 2 ^ (2 ^ n) + 1) :
+    ∃ k, p = k * 2 ^ (n + 2) + 1 := by
```

Comment:
Does this generalize to `p ∣ a ^ (2 ^ n) + 1 → ∃ k, p = k * 2 ^ (n + 1) + 1`? I think you can deduce the stronger `a = 2` version from this by applying the general result once to get `p % 8 = 1`, writing `(2 : ZMod p) = m ^ 2`, and reapplying the general result to `p ∣ m ^ (2 ^ (n + 1)) + 1`.

Reply:
Maybe. However, my real job strangles me positively at the moment. I'd prefer that this file be merged first, before I start going into this.

Reply:
It's probably worth including it in this PR since it shortens the proof (since you only need to make the `orderOf` argument once rather than twice) and generalizes the result.
```
lemma lem1 {a n p : ℕ} (hp : p.Prime) (hp2 : p ≠ 2) (hpdvd : p ∣ a ^ (2 ^ n) + 1) :
    ∃ k, p = k * 2 ^ (n + 1) + 1 := by
  have : Fact (2 < p) := Fact.mk (lt_of_le_of_ne hp.two_le hp2.symm)
  have : Fact p.Prime := Fact.mk hp
  have ha1 : (a : ZMod p) ^ (2 ^ n) = -1 := by
    rw [eq_neg_iff_add_eq_zero]
    exact_mod_cast (natCast_zmod_eq_zero_iff_dvd (a ^ (2 ^ n) + 1) p).mpr hpdvd
  have ha0 : (a : ZMod p) ≠ 0 := by
    intro h
    rw [h, zero_pow (pow_ne_zero n two_ne_zero), zero_eq_neg] at ha1
    exact one_ne_zero ha1
  have ha : orderOf (a : ZMod p) = 2 ^ (n + 1) := by
    apply orderOf_eq_prime_pow
    · rw [ha1]
      exact ZMod.neg_one_ne_one
    · rw [pow_succ, pow_mul, ha1, neg_one_sq]
  simpa [ha, dvd_def, Nat.sub_eq_iff_eq_add hp.one_le, mul_comm] using orderOf_dvd_card_sub_one ha0

lemma lem2 {n p : ℕ} (hn : 1 < n) (hp : p.Prime) (hpdvd : p ∣ 2 ^ (2 ^ n) + 1) :
    ∃ k, p = k * 2 ^ (n + 2) + 1 := by
  have : Fact p.Prime := Fact.mk hp
  have hp2 : p ≠ 2 := by
    exact ((even_pow.mpr ⟨even_two, pow_ne_zero n two_ne_zero⟩).add_one).ne_two_of_dvd_nat hpdvd
  have hp8 : p % 8 = 1 := by
    obtain ⟨k, rfl⟩ := lem1 hp hp2 hpdvd
    obtain ⟨n, rfl⟩ := Nat.exists_eq_add_of_le' hn
    rw [add_assoc, pow_add, ← mul_assoc, ← mod_add_mod, mul_mod]
    norm_num
  obtain ⟨a, ha⟩ := (exists_sq_eq_two_iff hp2).mpr (Or.inl hp8)
  suffices h : p ∣ a.val ^ (2 ^ (n + 1)) + 1 by
    exact lem1 hp hp2 h
  rw [← natCast_zmod_eq_zero_iff_dvd, Nat.cast_add, Nat.cast_one, Nat.cast_pow] at hpdvd ⊢
  rwa [ZMod.natCast_val, ZMod.cast_id, pow_succ', pow_mul, sq, ← ha]
```

Reply:
Great, thanks!